model/map-jpa/src/main/resources/META-INF/user-sessions/user/jpa-user-sessions-changelog-1.xml

```diff
@@ -0,0 +1,93 @@
+                <constraints nullable="false"/>
+            </column>
+            <column name="metadata" type="json"/>
+        </createTable>
+        <ext:addGeneratedColumn tableName="kc_user_session">
+            <ext:column name="entityversion" type="INTEGER" jsonColumn="metadata" jsonProperty="entityVersion"/>
+            <ext:column name="offline" type="BOOLEAN" jsonColumn="metadata" jsonProperty="fOffline"/>
+            <ext:column name="realmid" type="KC_KEY" jsonColumn="metadata" jsonProperty="fRealmId"/>
+            <ext:column name="userid" type="KC_KEY" jsonColumn="metadata" jsonProperty="fUserId"/>
+            <ext:column name="brokersessionid" type="KC_KEY" jsonColumn="metadata" jsonProperty="fBrokerSessionId"/>
+            <ext:column name="brokeruserid" type="KC_KEY" jsonColumn="metadata" jsonProperty="fBrokerUserId"/>
+            <ext:column name="lastsessionrefresh" type="BIGINT" jsonColumn="metadata" jsonProperty="fLastSessionRefresh"/>
+            <ext:column name="expiration" type="BIGINT" jsonColumn="metadata" jsonProperty="fExpiration"/>
+        </ext:addGeneratedColumn>
+        <createIndex tableName="kc_user_session" indexName="user_session_entityVersion">
+            <column name="entityversion"/>
+        </createIndex>
+        <createIndex tableName="kc_user_session" indexName="user_session_realm_offline_user">
+            <column name="realmid"/>
+            <column name="offline"/>
+            <column name="userid"/>
+        </createIndex>
+
+        <createTable tableName="kc_user_session_note">
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="fk_root" type="UUID">
+                <constraints foreignKeyName="user_session_note_fk_root_fkey" references="kc_user_session(id)" deleteCascade="true"/>
+            </column>
+            <column name="name" type="VARCHAR(255)"/>
+            <column name="value" type="TEXT"/>
+        </createTable>
+        <createIndex tableName="kc_user_session_note" indexName="kc_user_session_note_fk_root_name_key">
+            <column name="fk_root"/>
+            <column name="name"/>
+        </createIndex>
+
+        <createTable tableName="kc_user_session_client_session">
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="fk_root" type="UUID">
+                <constraints foreignKeyName="user_session_client_fk_root_fkey" references="kc_user_session(id)" deleteCascade="true"/>
+            </column>
+            <column name="name" type="VARCHAR(255)"/>
+            <column name="value" type="VARCHAR(255)"/>
+        </createTable>
+        <createIndex tableName="kc_user_session_client_session" indexName="user_session_client_fk_root_name">
+            <column name="fk_root"/>
+            <column name="name"/>
+        </createIndex>
```

Comment:
This table represents map of client session ids and client.id's. 

[`MapUserSessionEntity`](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionEntity.java#L113-L117) stores authenticated client session in format `Map<String, String>` where key is `clientUUID` and value is `clientSessionId`. `MapUserSessionAdapter` then loads client session on-by-one:  https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionAdapter.java#L116-L145. Because of that the table exist. 

Maybe we might consider to re-design this pattern. But that'd require another issue IMO.

Reply:
Can we eliminate this table by storing the client ID directly in the `kc_client_session` table?

Reply:
Client ID can be extracted from json and stored directly in `kc_client_session` table. 

I am wondering whether `JpaUserSessionEntity` should store rather `Set<JpaClientSessionEntity>` instead of `Set<JpaUserSessionClientSessionEntity>`. I am thinking about a workflow when all `JpaClientSessionEntity`s of the particular user session would be loaded from the database, then only `clientId` and `id` of it would be used to construct the `Map<String, String>` and then within `MapUserSessionAdapter` client sessions would be loaded by ids again. 

Shoudn't it be changed prior this?

Reply:
+1 to give `Set<JpaClientSessionEntity>` a try, and keeping my finger's crossed that orphan removal works as we would expect.

Reply:
(and yes, it could be its own issue as long as it is in phase 1)

Reply:
ftr: https://github.com/keycloak/keycloak/issues/12817 would be prerequisite for this.

Reply:
Tables restructured.